crates/ruff/src/rules/flake8_simplify/rules/ast_bool_op.rs

```diff
@@ -503,26 +504,28 @@ pub fn expr_or_not_expr(checker: &mut Checker, expr: &Expr) {
     }
 }
 
-pub fn is_short_circuit(
-    ctx: &Context,
-    expr: &Expr,
-    expected_op: &Boolop,
-) -> Option<(Location, Location)> {
+pub fn is_short_circuit(checker: &Checker, expr: &Expr, expected_op: &Boolop) -> Option<Edit> {
     let ExprKind::BoolOp { op, values, } = &expr.node else {
         return None;
     };
     if op != expected_op {
         return None;
     }
-    let short_circuit_value = match op {
-        Boolop::And => false,
-        Boolop::Or => true,
+    let short_circuit_truthiness = match op {
+        Boolop::And => Truthiness::Falsey,
+        Boolop::Or => Truthiness::Truthy,
     };
 
     let mut location = expr.location;
     for (value, next_value) in values.iter().tuple_windows() {
-        // Keep track of the location of the furthest-right, non-effectful expression.
-        if contains_effect(value, |id| ctx.is_builtin(id)) {
+        // Keep track of the location of the furthest-right, truthy or falsey expression
+        let value_truthiness = Truthiness::from_expr(value, |id| checker.ctx.is_builtin(id));
+        let next_value_truthiness =
+            Truthiness::from_expr(next_value, |id| checker.ctx.is_builtin(id));
+
+        // Keep track of the location of the furthest-right, non-effectful expression
+        if value_truthiness.is_unknown() && contains_effect(value, |id| checker.ctx.is_builtin(id))
```

Comment:
```python
a and ""
a or "a"
```
is replaced by
```python
""
"a"
```
This auto-fix would only work if the result of the boolean operations inside `bool(...)` or `if ...`.

We should only check for truthiness is unknown and not `contains_effect` in the other cases.

Reply:
I am working on a visitor to know if we are in a boolean context (i.e. boolean operation result will be converted to a boolean).